### PR TITLE
Charlesmchen/sae vs file types

### PR DESCRIPTION
**Pull request by charlesmchen**
* Adds support for sharing URLs in SAE.
  * Our entire SAE experience is premised around attachments, not text messages, so we'll share URLs (and other textual shares) as "oversize text message" attachments.
* Modified attachment approval view to properly preview "oversize text message" attachments.
* Fixed support for sharing arbitrary data (e.g. PDFs, .log, etc.)

PTAL @michaelkirk 

simulator screen shot - iphone x - 2018-01-11 at 13 32 43
simulator screen shot - iphone x - 2018-01-11 at 13 32 30

### CODE REVIEW

File: Signal/src/ViewControllers/ConversationView/ConversationViewItem.m [hunk 1]
```diff
@@ -309,30 +309,7 @@ - (DisplayableText *)displayableTextForInteractionId:(NSString *)interactionId
     DisplayableText *_Nullable displayableText = [[self displayableTextCache] objectForKey:interactionId];
     if (!displayableText) {
         NSString *text = textBlock();
-
-        // Only show up to N characters of text.
-        const NSUInteger kMaxTextDisplayLength = 1024;
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/ConversationView/ConversationViewItem.m]: Pulled this logic into `DisplayableText` so we can use it in the SAE.

File: SignalMessaging/attachments/MediaMessageView.swift [hunk 2]
```diff
@@ -277,6 +279,58 @@ public class MediaMessageView: UIView, OWSAudioAttachmentPlayerDelegate {
...
+        messageBubbleView.autoVCenterInSuperview()
+        messageBubbleView.autoHCenterInSuperview()
+        messageBubbleView.autoPinEdge(toSuperviewEdge: .leading, withInset: 25, relation: .greaterThanOrEqual)
+        messageBubbleView.autoPinEdge(toSuperviewEdge: .trailing, withInset: 25, relation: .greaterThanOrEqual)
+
+        messageBubbleView.addSubview(messageTextView)
+        messageTextView.autoPinTopToSuperview(withMargin:10)
+        messageTextView.autoPinBottomToSuperview(withMargin:10)
+        messageTextView.autoPinLeadingToSuperview(withMargin:10)
+        messageTextView.autoPinTrailingToSuperview(withMargin:15)
```
Comment (charlesmchen) [inline SignalMessaging/attachments/MediaMessageView.swift]: As discussed, I'd like to rework bubble appearance soon.  When we do that, I'll pull out these constants to somewhere in `SignalMessaging`.

File: SignalMessaging/categories/UIView+OWS.h [hunk 3]
```diff
@@ -83,6 +83,12 @@ CGFloat ScaleFromIPhone5(CGFloat iPhone5Value);
 - (NSLayoutConstraint *)autoPinLeadingToSuperviewWithMargin:(CGFloat)margin;
 - (NSLayoutConstraint *)autoPinTrailingToSuperview;
 - (NSLayoutConstraint *)autoPinTrailingToSuperviewWithMargin:(CGFloat)margin;
+
+- (NSLayoutConstraint *)autoPinTopToSuperview;
+- (NSLayoutConstraint *)autoPinTopToSuperviewWithMargin:(CGFloat)margin;
+- (NSLayoutConstraint *)autoPinBottomToSuperview;
+- (NSLayoutConstraint *)autoPinBottomToSuperviewWithMargin:(CGFloat)margin;
```
Comment (charlesmchen) [inline SignalMessaging/categories/UIView+OWS.h]: Picked this over from hotfix and elaborated a bit.

File: SignalMessaging/categories/UIView+OWS.m [hunk 4]
```diff
@@ -268,6 +268,8 @@ - (NSLayoutConstraint *)autoPinLeadingToSuperview
 
 - (NSLayoutConstraint *)autoPinLeadingToSuperviewWithMargin:(CGFloat)margin
 {
+    self.translatesAutoresizingMaskIntoConstraints = NO;
```
Comment (charlesmchen) [inline SignalMessaging/categories/UIView+OWS.m]: We always want to disable `translatesAutoresizingMaskIntoConstraints` whenever we start using auto layout constraints to pin edges of the view.

**Comment by charlesmchen** [comment]
@michaelkirk Please hold off on reviewing this PR.  I'm going to update it to reflect our conversation at standup.

**Comment by charlesmchen** [comment]
PTAL @michaelkirk 

I've made a number of modifications:

* Refine filtering of share types using a query.
* Handle data-based share item providers.
   * Some share items' providers are an instance of  `NSData`, not an `NSURL`.  We can write those to a temporary file.
* Handle of contacts, but filter them out.  We can turn on contact sharing in a subsequent release.
* Unpack oversize text messages if possible.  This simplifies sharing of URLs, and other textual shares.
* Don't add captions to text messages.

File: SignalShareExtension/ShareViewController.swift [hunk 5]
```diff
@@ -544,7 +639,20 @@ public class ShareViewController: UINavigationController, ShareViewDelegate, SAE
     // Perhaps the AVFoundation APIs require some extra file system permssion we don't have in the
     // passed through URL.
     private func isVideoNeedingRelocation(itemProvider: NSItemProvider, itemUrl: URL) -> Bool {
-        guard MIMETypeUtil.utiType(forFileExtension: itemUrl.pathExtension) == kUTTypeMPEG4 as String else {
+        if ShareViewController.isSpecialItem(itemProvider:itemProvider) {
```
Comment (michaelkirk) [inline SignalShareExtension/ShareViewController.swift]: Do these extra 12 lines add any value? It seems redundant and just muddies the water, but maybe there's an edge case I don't understand.

**Comment by charlesmchen** [comment]
>I think ultimately we want people to be able to caption URLs, I'm generally OK with incremental progress, but do you have a vision for how we'd get there from here?

Seems like we have the following options:

* Use an existing file format for transmitting URLs, e.g. `.webloc` or `.url`.  My understanding is that neither format is well-supported.  
   * `.webloc` is Apple-specific.
   * See: http://www.lyberty.com/encyc/articles/tech/dot_url_format_-_an_unofficial_guide.html
* Define a Signal-specific attachment format used only internally for communications between Signal clients (e.g. UTF-8 representation of the URL, send with a custom mimetype).  The other clients would have to be modified to support it.
* Send URLs as oversize text attachments + caption, but again we'd need to modify all clients to present these text+text messages properly.  
* Send URLs as text messages without a caption (as I have it here).
* Place the URL in the caption field and let users edit it & add extra text around it (WhatsApp does this). As you noted, we'd probably want to change the appearance of the "attachment approval".  This would also require some other work, since the share logic is all attachment-oriented at the moment.  E.g. we could use a custom "unsent URL" mimetype, which the attachment approval and sending logic would special-case.
* Show the caption field + simply append it (with a separating whitespace?) to the URL.

Very happy to switch to a different option - let me know.

File: SignalMessaging/attachments/SharingThreadPickerViewController.m [hunk 6]
```diff
@@ -193,39 +193,63 @@ - (void)tryToSendAttachment:(SignalAttachment *)attachment fromViewController:(U
...
-
-                                       self.outgoingMessage = outgoingMessage;
+    [fromViewController
+        presentViewController:progressAlert
+                     animated:YES
+                   completion:^(void) {
+                       __block TSOutgoingMessage *outgoingMessage = nil;
+                       if (self.attachment.isOversizeText
+                           && self.attachment.dataLength <= kOversizeTextMessageSizeThreshold) {
+                           // Try to unpack oversize text messages and send them as regular
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalMessaging/attachments/SharingThreadPickerViewController.m]: This has been removed.  The goal was expediency and pragmatism, but we have now added a separate view & code path for textual shares.

File: SignalMessaging/categories/UIView+OWS.m [hunk 7]
```diff
@@ -299,6 +303,44 @@ - (NSLayoutConstraint *)autoPinTrailingToSuperviewWithMargin:(CGFloat)margin
...
+        NSLayoutConstraint *constraint =
+            [self.bottomAnchor constraintEqualToAnchor:self.superview.layoutMarginsGuide.bottomAnchor constant:-margin];
+        constraint.active = YES;
+        return constraint;
+    } else {
+        return [self autoPinEdgeToSuperviewEdge:ALEdgeBottom withInset:margin];
+    }
+}
+
+- (NSLayoutConstraint *)autoPinTopToSuperview
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalMessaging/categories/UIView+OWS.m]: That seems like a good idea.  Please make a card?

File: SignalShareExtension/Info.plist [hunk 8]
```diff
@@ -53,21 +53,21 @@
...
+            <key>NSExtensionActivationRule</key>
+            <string>SUBQUERY (
+                extensionItems,
+                $extensionItem,
+                SUBQUERY (
+                $extensionItem.attachments,
+                $attachment,
+                (
+                ANY $attachment.registeredTypeIdentifiers UTI-CONFORMS-TO "public.data"
+                || ANY $attachment.registeredTypeIdentifiers UTI-CONFORMS-TO "public.url"
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalShareExtension/Info.plist]: I'm still hoping that we can accept all public.data, but we need to test edge cases to make sure this is true.

File: SignalShareExtension/ShareViewController.swift [hunk 9]
```diff
@@ -73,7 +73,7 @@ public class ShareViewController: UINavigationController, ShareViewDelegate, SAE
         self.loadViewController = loadViewController
 
         // Don't display load screen immediately, in hopes that we can avoid it altogether.
-        after(seconds: 0.5).then { () -> Void in
+        after(seconds: 0.25).then { () -> Void in
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalShareExtension/ShareViewController.swift]: Reverted.

File: SignalShareExtension/ShareViewController.swift [hunk 10]
```diff
@@ -455,13 +499,44 @@ public class ShareViewController: UINavigationController, ShareViewDelegate, SAE
...
+            guard let provider = provider else {
+                let missingProviderError = ShareViewControllerError.assertionError(description: "missing item provider")
+                reject(missingProviderError)
                 return
             }
 
-            fulfill(url)
+            Logger.info("\(self.logTag) provider type: \(type(of:provider))")
+
+            if let data = provider as? Data {
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalShareExtension/ShareViewController.swift]: Yes, it's "many things".  Some file shares will be received as (file) URLs and others as `Data`.  `NSItemProvider` offers no guarantees and the new API methods that afford more control over this behavior a) don't always work as expected and b) require iOS 11 making them useless for us.

File: SignalShareExtension/ShareViewController.swift [hunk 11]
```diff
@@ -455,13 +499,44 @@ public class ShareViewController: UINavigationController, ShareViewDelegate, SAE
...
+                var tempFileName = NSUUID().uuidString
+                if let customFileExtension = MIMETypeUtil.fileExtension(forUTIType:utiType) {
+                    tempFileName += "." + customFileExtension
+                }
+                let tempFilePath = (tempDirPath as NSString).appendingPathComponent(tempFileName)
+                do {
+                    let fileUrl = URL(fileURLWithPath:tempFilePath)
+                    try data.write(to: fileUrl)
+
+                    // Don't back up Giphy downloads.
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalShareExtension/ShareViewController.swift]: Great idea.

[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalShareExtension/ShareViewController.swift]: Done.

File: SignalShareExtension/ShareViewController.swift (hunk 5 above)
Comment (charlesmchen) [reply SignalShareExtension/ShareViewController.swift]: I've streamlined this a bit, but yes - `MIMETypeUtil.utiType(forFileExtension:...)` isn't safe to call for URLs/paths without a file extension.

**Comment by charlesmchen** [comment]
PTAL @michaelkirk 

I've made yet more changes:

* Improve handling of "textual" shares, e.g. text snippets and URLs.
* Added a "message approval view" to complement "attachment approval view".

More comments in-line.

### Group Recipient, Text Snippet

simulator screen shot - iphone x - 2018-01-17 at 15 55 28

### Single Recipient (with optional profile name), URL.

simulator screen shot - iphone x - 2018-01-17 at 15 53 25

File: Signal.xcodeproj/project.pbxproj [hunk 12]
```diff
@@ -42,7 +42,12 @@
 		344F248720069ECB00CFB4F4 /* ModalActivityIndicatorViewController.swift in Sources */ = {isa = PBXBuildFile; fileRef = 344F248620069ECB00CFB4F4 /* ModalActivityIndicatorViewController.swift */; };
 		344F248A20069F0600CFB4F4 /* ViewControllerUtils.h in Headers */ = {isa = PBXBuildFile; fileRef = 344F248820069F0600CFB4F4 /* ViewControllerUtils.h */; };
 		344F248B20069F0600CFB4F4 /* ViewControllerUtils.m in Sources */ = {isa = PBXBuildFile; fileRef = 344F248920069F0600CFB4F4 /* ViewControllerUtils.m */; };
-		3456D2C41FFFCC70001EA55D /* OWSBackupViewController.m in Sources */ = {isa = PBXBuildFile; fileRef = 3456D2C21FFFCC6F001EA55D /* OWSBackupViewController.m */; };
+		344F248D2007CCD600CFB4F4 /* DisplayableText.swift in Sources */ = {isa = PBXBuildFile; fileRef = 344F248C2007CCD600CFB4F4 /* DisplayableText.swift */; };
```
Comment (charlesmchen) [inline Signal.xcodeproj/project.pbxproj]: We could move some of these files back from `SignalMessaging` to `Signal`, but I'm inclined to leave them in their new locations.  I could see using them again in the future.

File: SignalMessaging/attachments/MessageApprovalViewController.swift [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+@objc
+public protocol MessageApprovalViewControllerDelegate: class {
+    func messageApproval(_ messageApproval: MessageApprovalViewController, didApproveMessage messageText: String)
+    func messageApprovalDidCancel(_ messageApproval: MessageApprovalViewController)
+}
+
+@objc
+public class MessageApprovalViewController: OWSViewController, UITextViewDelegate {
+
+    let TAG = "[MessageApprovalViewController]"
```
Comment (charlesmchen) [inline SignalMessaging/attachments/MessageApprovalViewController.swift]: I tried to conform to the arrangement of the attachment approval view where possible.

File: SignalMessaging/attachments/MessageApprovalViewController.swift [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+                                                      comment: "Title for the 'message approval' dialog.")
+    }
+
+    private func updateToolbar() {
+        var items = [UIBarButtonItem]()
+
+        let cancelButton = UIBarButtonItem(barButtonSystemItem: .stop, target: self, action: #selector(cancelPressed))
+        items.append(cancelButton)
+
+        if textView.text.count > 0 {
```
Comment (charlesmchen) [inline SignalMessaging/attachments/MessageApprovalViewController.swift]: Don't let users send empty text messages.

File: SignalMessaging/attachments/MessageApprovalViewController.swift [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+            return recipientRow
+        }
+
+        nameLabel.attributedText = contactsManager.formattedFullName(forRecipientId:contactThread.contactIdentifier(), font:font)
+        nameLabel.textColor = UIColor.black
+
+        if let profileName = self.profileName(contactThread:contactThread) {
+            // If there's a profile name worth showing, add it as a second line below the name.
+            let profileNameLabel = UILabel()
+            profileNameLabel.textColor = UIColor.ows_darkGray
```
Comment (charlesmchen) [inline SignalMessaging/attachments/MessageApprovalViewController.swift]: It's not a pixel-for-pixel match, but I've followed the pattern that we used in `ContactTableViewCell` of showing the optional "profile name" below the "name", with a lighter color.

File: SignalShareExtension/ShareViewController.swift [hunk 16]
```diff
@@ -428,46 +498,109 @@ public class ShareViewController: UINavigationController, ShareViewDelegate, SAE
...
-        guard let utiType = matchingUtiType else {
+        // We need to be very careful about which UTI type we use.
+        //
+        // * In the case of "textual" shares (e.g. web URLs and text snippets), we want to
+        //   coerce the UTI type to kUTTypeURL or kUTTypeText.
+        // * We want to treat shared files as file attachments.  Therefore we do not
+        //   want to treat file URLs like web URLs.
+        // * UTIs aren't very descriptive (there are far more MIME types than UTI types)
+        //   so in the case of file attachments we try to refine the attachment type
+        //   using the file extension.
```
Comment (charlesmchen) [inline SignalShareExtension/ShareViewController.swift]: This comment is important.

File: SignalShareExtension/ShareViewController.swift [hunk 17]
```diff
@@ -428,46 +498,109 @@ public class ShareViewController: UINavigationController, ShareViewDelegate, SAE
...
+        // since it can yield values of arbitrary type.  It has a highly unusual design
+        // in which its behavior depends on the _type_ of the completion handler.
+        // loadItem(forTypeIdentifier:...) tries to satisfy the expected type of the
+        // completion handler.  This "hinting" only works in Objective-C.  In Swift,
+        // The type of the completion handler must agree with the param type.
+        //
+        // Unfortunately, we have no alternative.  Therefore, we face the real possibility
+        // of receiving an "unexpected type" that we don't know how to handle.
+        //
+        // See: https://developer.apple.com/documentation/foundation/nsitemprovider/1403900-loaditemfortypeidentifier
```
Comment (charlesmchen) [inline SignalShareExtension/ShareViewController.swift]: This comment is important.

File: SignalShareExtension/ShareViewController.swift [hunk 18]
```diff
@@ -249,7 +249,10 @@ public class ShareViewController: UINavigationController, ShareViewDelegate, SAE
             // is has already saved their local profile key in the main app.
             showNotReadyView()
         } else {
-            presentConversationPicker()
+            DispatchQueue.main.async { [weak self] in
```
Comment (charlesmchen) [inline SignalShareExtension/ShareViewController.swift]: `presentConversationPicker` can be expensive, so I've modified this to call it async (still on `main`).  That should help ensure that loading the share view controller occurs in a timely way.

File: SignalMessaging/categories/UIView+OWS.m (hunk 7 above)
[the remark this answers is not in the document]
Comment (michaelkirk) [reply SignalMessaging/categories/UIView+OWS.m]: Added.

File: SignalMessaging/attachments/MessageApprovalViewController.swift (hunk 15 above)
Comment (michaelkirk) [reply SignalMessaging/attachments/MessageApprovalViewController.swift]: We're eventually going to allow multiple recipients, right?

In which case we'll potentially need to wrap the list of names to multiple lines. I think in that context having the profile name below the phone number will be strange, and it would be preferable for the profile name to be on the same line as the phone number:

ASCII mockup:

e.g. wrapping a long line of recipients:

```
To: +1-555-555-5555 ~Alice, Bob,
Charles, Danielle 
```

A little awkward when only some recipients have profile names.

```
To: +1-555-555-5555, Bob,
    ~Alice
Charles, Danielle 
```

I guess I'm envisioning it would look like a `To:` line in any other messaging/email app.

I'm ok with the current approach if you'd rather not do that work now.

File: SignalMessaging/attachments/SharingThreadPickerViewController.m [hunk 19]
```diff
@@ -158,15 +200,43 @@ - (void)attachmentApproval:(AttachmentApprovalViewController *)attachmentApprova
...
+                 fromViewController:approvalViewController];
+}
+
+- (void)messageApprovalDidCancel:(MessageApprovalViewController *)approvalViewController
+{
+    [self cancelShareExperience];
+}
+
 #pragma mark - Helpers
+typedef void (^SendMessageBlock)(SendCompletionBlock completion);
```
Comment (michaelkirk) [inline SignalMessaging/attachments/SharingThreadPickerViewController.m]: this is redundant with the declaration at the top of the file

File: SignalMessaging/attachments/SharingThreadPickerViewController.m [hunk 20]
```diff
@@ -193,38 +263,30 @@ - (void)tryToSendAttachment:(SignalAttachment *)attachment fromViewController:(U
...
                                    completion:^(void) {
-                                       __block TSOutgoingMessage *outgoingMessage =
-                                           [ThreadUtil sendMessageWithAttachment:self.attachment
-                                                                        inThread:self.thread
-                                                                   messageSender:self.messageSender
-                                                                      completion:^(NSError *_Nullable error) {
-                                                                          sendCompletion(error, outgoingMessage);
-                                                                      }];
-
-                                       self.outgoingMessage = outgoingMessage;
```
Comment (michaelkirk) [inline SignalMessaging/attachments/SharingThreadPickerViewController.m]: It seems like we lost this assignment which is required to show upload progress.

File: SignalMessaging/categories/UIView+OWS.h (hunk 3 above)
Comment (michaelkirk) [reply SignalMessaging/categories/UIView+OWS.h]: I think these are now in master, I think this branch needs a rebase.

File: SignalShareExtension/ShareViewController.swift [hunk 21]
```diff
@@ -428,46 +498,103 @@ public class ShareViewController: UINavigationController, ShareViewDelegate, SAE
...
+        //   using the file extension.
+        guard let srcUtiType = ShareViewController.utiTypeForItem(itemProvider: itemProvider) else {
             let error = ShareViewControllerError.unsupportedMedia
             return Promise(error: error)
         }
-        Logger.debug("\(logTag) matched utiType: \(utiType)")
+        Logger.debug("\(logTag) matched utiType: \(srcUtiType)")
 
-        let (promise, fulfill, reject) = Promise<URL>.pending()
+        let (promise, fulfill, reject) = Promise<(URL, String)>.pending()
```
Comment (michaelkirk) [inline SignalShareExtension/ShareViewController.swift]: TIOLI: Now that this method is longer and this promise returns multiple objects, it would help with readbility to return a named tuple.


e.g. 
```
let (promise, fulfill, reject) = Promise<(attachmentURL: URL, utiType: String)>.pending()
```

File: SignalShareExtension/ShareViewController.swift [hunk 22]
```diff
@@ -477,16 +604,19 @@ public class ShareViewController: UINavigationController, ShareViewDelegate, SAE
                 }
             }()
 
-            Logger.debug("\(self.logTag) building DataSource with url: \(url)")
+            Logger.debug("\(self.logTag) building DataSource with url: \(url), utiType: \(utiType)")
 
-            guard let dataSource = DataSourcePath.dataSource(with: url) else {
+            guard let dataSource = ShareViewController.createDataSource(utiType : utiType, url : url, customFileName : customFileName) else {
```
Comment (michaelkirk) [inline SignalShareExtension/ShareViewController.swift]: nit: param format

File: SignalMessaging/categories/UIView+OWS.h (hunk 3 above)
Comment (charlesmchen) [reply SignalMessaging/categories/UIView+OWS.h]: Yup, done.

File: SignalMessaging/attachments/MessageApprovalViewController.swift (hunk 15 above)
Comment (charlesmchen) [reply SignalMessaging/attachments/MessageApprovalViewController.swift]: Great points.  I hesitated to combine them onto a single line for a few reasons.  One is to feel consistent with how these names are presented in the main app (currently).  Another is because I seem to recall security concerns around ambiguity between where one ended and another begun (which could be mitigated by font, font weight, font size, color, etc.).

Let's revisit this when we do the work to allow multiple recipients.

File: SignalMessaging/attachments/MessageApprovalViewController.swift [hunk 23]
```diff
@@ -0,0 +1,231 @@
...
+        topBorder.autoSetDimension(.height, toSize: borderThickness)
+
+        let bottomBorder = UIView.container()
+        bottomBorder.backgroundColor = borderColor
+        recipientRow.addSubview(bottomBorder)
+        bottomBorder.autoPinWidthToSuperview()
+        bottomBorder.autoPinBottomToSuperview()
+        bottomBorder.autoSetDimension(.height, toSize: borderThickness)
+
+        guard let font = UIFont.ows_regularFont(withSize:ScaleFromIPhone5To7Plus(14.0, 18.0)) else {
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply SignalMessaging/attachments/MessageApprovalViewController.swift]: Done.

File: SignalMessaging/attachments/SharingThreadPickerViewController.m (hunk 19 above)
Comment (charlesmchen) [reply SignalMessaging/attachments/SharingThreadPickerViewController.m]: Removed.

File: SignalMessaging/attachments/SharingThreadPickerViewController.m (hunk 20 above)
Comment (charlesmchen) [reply SignalMessaging/attachments/SharingThreadPickerViewController.m]: Ah, good catch.  Fixed and added a comment to make clear why this is necessary.

File: SignalShareExtension/ShareViewController.swift (hunk 21 above)
Comment (charlesmchen) [reply SignalShareExtension/ShareViewController.swift]: Done.

File: SignalShareExtension/ShareViewController.swift (hunk 22 above)
Comment (charlesmchen) [reply SignalShareExtension/ShareViewController.swift]: Done.
